tests/drivers/spi/qspi/src/main.c

```diff
@@ -0,0 +1,375 @@
+}
+
+/**
+ * @brief Write data into flash using spi api
+ * @details
+ * - flash write enable
+ * - write data into flash using spi api
+ */
+void test_spi_single_write(void)
+{
+	int ret;
+
+	/* write enable */
+	safbuf[0] = SPI_WRITE_ENABLE;
+	txb.buf = &safbuf;
+	txb.len = 1U;
+
+	tx_bufs.buffers = (const struct spi_buf *)&txb;
+	tx_bufs.count = 1U;
+
+	rx_bufs.buffers = NULL;
+	rx_bufs.count = 0U;
+
+	ret = spi_transceive(spi_dev,
+			(const struct spi_config *)&spi_cfg_single,
+			(const struct spi_buf_set *)&tx_bufs,
+			(const struct spi_buf_set *)&rx_bufs);
+	zassert_true(ret == 0, "Send write enable spi_transceive failure: "
+			"error %d", ret);
+
+	/* write data start from address 0x000010 */
+	safbuf[0] = SPI_WRITE_DATA;
+	safbuf[1] = 0x10U;
+	safbuf[2] = 0x0U;
+	safbuf[3] = 0x0U;
+	memcpy(&safbuf[4], buffer_tx, BUF_SIZE);
+
+	txb.buf = &safbuf;
+	txb.len = 4 + BUF_SIZE;
+
+	tx_bufs.buffers = (const struct spi_buf *)&txb;
+	tx_bufs.count = 1U;
+
+	rx_bufs.buffers = NULL;
+	rx_bufs.count = 0U;
+
+	ret = spi_transceive(spi_dev,
+			(const struct spi_config *)&spi_cfg_single,
+			(const struct spi_buf_set *)&tx_bufs,
+			(const struct spi_buf_set *)&rx_bufs);
+	zassert_true(ret == 0, "Send write data spi_transceive failure: "
+			"error %d", ret);
+}
+
+/**
+ * @brief Read data from flash using spi single mode
+ * @details
+ * - read data using spi single mode
+ * - check read buffer data whether correct
+ */
+void test_spi_single_read(void)
+{
+	int ret;
+
+	/* read data using spi single mode */
+	safbuf[0] = SPI_FAST_READ_DATA;
+	safbuf[1] = 0x10U;
+	safbuf[2] = 0x0U;
+	safbuf[3] = 0x0U;
+	safbuf[4] = 0x0U;
+
+	txb.buf = &safbuf;
+	txb.len = 5U;
+
+	tx_bufs.buffers = (const struct spi_buf *)&txb;
+	tx_bufs.count = 1U;
+
+	memset(safbuf2, 0, BUF_SIZE);
+	rxb.buf = &safbuf2;
+	rxb.len = BUF_SIZE;
+
+	rx_bufs.buffers = (const struct spi_buf *)&rxb;
+	rx_bufs.count = 1U;
+
+	ret = spi_transceive(spi_dev,
+			(const struct spi_config *)&spi_cfg_single,
+			(const struct spi_buf_set *)&tx_bufs,
+			(const struct spi_buf_set *)&rx_bufs);
+	zassert_true(ret == 0, "Send fast read data spi_transceive failure: "
+			"error %d", ret);
+
+	/* check read buffer data whether correct */
+	zassert_true(memcmp(buffer_tx, safbuf2, BUF_SIZE) == 0,
+			"Buffer read data is different to write data");
+}
+
+
+/**
+ * @brief Read data from flash using spi dual mode
+ * @details
+ * - read data using spi dual mode
+ * - check read buffer data whether correct
+ */
+void test_spi_dual_read(void)
+{
+	int ret;
+
+	/* configure spi dual mode */
+	spi_cfg_dual.frequency = TEST_FREQ_HZ;
+	spi_cfg_dual.operation = SPI_OP_MODE_MASTER | SPI_TRANSFER_MSB
+		| SPI_WORD_SET(8) | SPI_LINES_DUAL;
+	spi_cfg_dual.slave = 0;
+	spi_cfg_dual.cs = NULL;
+
+	/* read data using spi dual mode */
+	safbuf[0] = SPI_DUAL_FAST_READ_DATA;
+	safbuf[1] = 0x10U;
+	safbuf[2] = 0x0U;
+	safbuf[3] = 0x0U;
+	safbuf[4] = 0x0U;
```

Comment:
Using transmit of 0 for dummy clocks can result in the controller and target both driving the pins at some point during the dummy clocks. This doesn't matter for single(full-duplex) mode because transmit and receive are on different pins. The driver provides for dummy clock generation using another struct spi_buf with buf=NULL and len=number of bytes that will generate the correct number of dummy clocks. 8 clocks * 2 bits/clock = 16 bits or 2 bytes.

Reply:
I have changed to use spi_buf to set dummy clocks.